haystack/nodes/base.py

```diff
@@ -61,14 +61,25 @@ class BaseComponent(ABC):
 
     def __init__(self):
         # a small subset of the component's parameters is sent in an event after applying filters defined in haystack.telemetry.NonPrivateParameters
-        send_custom_event(event=f"{type(self).__name__} initialized", payload=self._component_config.get("params", {}))
+        component_params = self._component_config.get("params", {})
+        send_custom_event(event=f"{type(self).__name__} initialized", payload=component_params)
+        self.outgoing_edges = self._calculate_outgoing_edges(component_params=component_params)
```

Comment:
Dumb question, but is `self.outgoing_edges` or even `cls.outgoing_edges` ever used anymore? The pipeline was checking these attributes, but is now using `_calculate_outgoing_edges` directly, so... where else are these needed? The only place where I see it used is the check you implemented a few lines below :smile: Maybe we don't need them anymore.

Reply:
Yes it's technically not being used anymore. However we need some place to store the class level default value which is used by `_calculate_outgoing_edges` as default. I think especially for custom nodes that don't need a dynamic value it's handy to define that with `outgoing_edges`.

Reply:
An alternative would be to drop `outgoing_edges` and make `_calvulate_outgoing_edges` mandatory to implement for subclasses by making it abstract. That however would be a breaking change (again especially for custom nodes). And given that dynamic values were not possible till now, the gain for our users would be zero, but the pain not.

Reply:
What I'm mostly worried about here is having two difference sources of truth for `outgoing_edges`. It can only confuse users later on. 

An approach could be to still remove `outgoing_edges`, and then:

1. Change `_calculate_outgoing_edges` into `get_outgoing_edges` 

2. Implement `get_outgoing_edges` in `BaseComponent` and make it return 1. The majority of nodes has `outgoing_edges=1` anyway, so this would actually simplify them.

3. Nodes that have a larger number of outgoing edges (I think now it's like 2 or 3 nodes in total) will override `get_outgoing_edges` to return whatever value they need.

This gives us one single source of truth while keeping the flexibility and without asking anyone to change their custom nodes. I'd maybe add a warning that tells users to use `get_outgoing_edges` if it detects that `cls.outgoing_edges` is defined, and log an error if `cls.outgoing_edges > 1`, but that's it.

Do you see any drawback?

Reply:
Drawback of that solution would be:
- documentation needs to be updated: see https://haystack.deepset.ai/pipeline_nodes/custom-nodes
- tutorial 11 needs to be updated
- we have a breaking change as `outgoing_edges` is not working as before
- we have no way to force people to select an outgoing_edges value at implementation/compile time: if people didn't override `get_outgoing_edges` appropriately for multiple outgoing edges they might not understand what they have to do without looking into the docs: the error message then would need to address both cases. Either they added too many outgoing edges or the didn't override `get_outgoing_edges` appropriately.

Reply:
You're right, overall it's probably better to maintain some backward compatibility here. After all this is a change that will affect few nodes, it's better if it stays this way :+1:

I'll keep the breaking changes for the bigger pipelines refactoring :smile: